ff-parser/src/main/java/org/uniprot/core/flatfile/tool/ca/DefaultCatalyticActivityValidator.java

```diff
@@ -0,0 +1,127 @@
+            }
+        }
+        Optional<CrossReference<ReactionDatabase>> rheaRef =
+                reaction.getReactionCrossReferences().stream()
+                        .filter(
+                                val ->
+                                        (val.getDatabase().equals(ReactionDatabase.RHEA))
+                                                && val.getId().startsWith(RHEA_PREFIX))
+                        .findFirst();
+        if (!rheaRef.isPresent()) {
+            return Optional.empty();
+        }
+        CatalyticActivity ca = repository.getByRheaId(rheaRef.get().getId());
+        if (ca == null) {
+            return Optional.empty();
+        }
+
+        CatalyticActivityCommentBuilder builder = new CatalyticActivityCommentBuilder();
+        builder.reaction(transformReaction(reaction, ca, rheaRef.get()));
+
+        List<PhysiologicalReaction> physioReactions =
+                comment.getPhysiologicalReactions().stream()
+                        .map(val -> transformPhysiologicalReaction(ca, val))
+                        .filter(val -> val != null)
+                        .collect(Collectors.toList());
+        builder.physiologicalReactionsSet(physioReactions);
+
+        return Optional.ofNullable(builder.build());
+    }
+
+    private Reaction transformReaction(
+            Reaction reaction, CatalyticActivity ca, CrossReference<ReactionDatabase> rhea) {
+        ReactionBuilder builder = new ReactionBuilder();
+        builder.name(ca.getText());
+        List<CrossReference<ReactionDatabase>> xrefs = new ArrayList<>();
+        xrefs.add(createReactRef(rhea.getDatabase(), rhea.getId()));
+        ca.getReactantIds().stream().map(this::parseReactantId).forEach(val -> xrefs.add(val));
+        builder.reactionCrossReferencesSet(xrefs);
+
+        if (Utils.notNull(reaction.getEcNumber())) {
+            if (ca.getEcs().contains(reaction.getEcNumber().getValue())) {
+                builder.ecNumber(reaction.getEcNumber().getValue());
+            }
+        }
+        builder.evidencesSet(reaction.getEvidences());
+        return builder.build();
+    }
+
+    private PhysiologicalReaction transformPhysiologicalReaction(
+            CatalyticActivity ca, PhysiologicalReaction physioReaction) {
+        CrossReference<ReactionDatabase> xref = physioReaction.getReactionCrossReference();
+        if ((xref == null) || !xref.getDatabase().equals(ReactionDatabase.RHEA)) {
+            return null;
+        }
+        PhysiologicalDirectionType type = physioReaction.getDirectionType();
+        if (type == PhysiologicalDirectionType.LEFT_TO_RIGHT) {
+            if (xref.getId().equals(ca.getRheaLr())) return copy(physioReaction);
+        } else if (type == PhysiologicalDirectionType.RIGHT_TO_LEFT) {
+            if (xref.getId().equals(ca.getRheaRl())) return copy(physioReaction);
+        }
+        return null;
+    }
+
+    private PhysiologicalReaction copy(PhysiologicalReaction data) {
```

Comment:
Optional: I think we might be able to use PhysiologicalReactionBuilder.from method, and replace reactionCrossReferences (if necessary). This is other way to do it.. not sure if it is worst the change...

Reply:
good point. In fact, no copy is required. I will make a copy anyway